internal/tofu/transform_targets.go

```diff
@@ -73,10 +81,115 @@ func (t *TargetsTransformer) selectTargetedNodes(g *Graph, addrs []addrs.Targeta
+	for _, v := range vertices {
+		vertexAddr := t.getTargetableNodeResourceAddr(v)
+		if vertexAddr == nil {
+			continue
+		}
+
+		targetableNodes.Add(v)
+
+		nodeExcluded := t.nodeIsExcluded(vertexAddr, excludes)
+		if nodeExcluded {
+			excludedNodes.Add(v)
+		}
+
+		if nodeExcluded || t.nodeDescendantsExcluded(vertexAddr, excludes) {
+			deps, _ := g.Descendents(v)
+			for _, d := range deps {
+				// In general, we'd like to exclude any descendant targetable node of the current node.
+				// We exclude any resource dependent on this resource (which is more general than resources dependent
+				// on the resource instance, but is in-line with how -target works).
+				//
+				// The exception to this is when excluding a specific instance of a resource that has multiple instances.
+				// During apply, the specific instance tofu.NodeApplyableResourceInstance would be dependent on the
+				// resource tofu.nodeExpandApplyableResource.
+				// Since we do not want to exclude all resource instances (other than the ones that we've explicitly
+				// excluded), we should only exclude dependents whose target is not contained in the current node.
+				depVertexAddr := t.getTargetableNodeResourceAddr(d)
+				if depVertexAddr != nil && !vertexAddr.TargetContains(depVertexAddr) {
+					excludedNodes.Add(d)
+				}
+			}
+		}
+	}
+
+	// Step 2: Of the targetable nodes that were not excluded, build the graph similarly to -target
+	for _, v := range targetableNodes {
+		if !excludedNodes.Include(v) {
+			targetedNodes.Add(v)
+
+			// We inform nodes that ask about the list of excludes - helps for nodes
+			// that need to dynamically expand. Note that this only occurs for nodes
+			// that are targetable and we didn't exclude
+			if tn, ok := v.(GraphNodeTargetable); ok {
+				tn.SetExcludes(excludes)
+			}
+
+			deps, _ := g.Ancestors(v)
+			for _, d := range deps {
+				targetedNodes.Add(d)
+			}
+		}
+	}
+
+	// Step 3: Add outputs
+	targetedOutputNodes := t.getTargetedOutputNodes(targetedNodes, g)
+	for _, outputNode := range targetedOutputNodes {
+		targetedNodes.Add(outputNode)
+	}
+
+	return targetedNodes
+}
+
+func (t *TargetsTransformer) getTargetedOutputNodes(targetedNodes dag.Set, graph *Graph) dag.Set {
 	// It is expected that outputs which are only derived from targeted
 	// resources are also updated. While we don't include any other possible
 	// side effects from the targeted nodes, these are added because outputs
 	// cannot be targeted on their own.
+	//
+	// Note: This behaviour has some quirks, as there are specific cases where
+	// you would think an output should not be updated, but it is
+	// For example, when there's a module call with an input that is dependent
+	// on a root resource, and only the root resource is targeted, any output
+	// that depends on a module output might be updated, if said module output
+	// does not depend on any resource of the module itself.
+	// Right now, we will not change this behaviour
```

Comment:
can you explain in a comment here about why we're not doing this, and when we could do this please?

Reply:
0e8d58db30ff3ad339b1d8dd409847dd9eff6c9d